test/e2e/framework/internal/unittests/testapiserver.go

```diff
@@ -0,0 +1,131 @@
+*/
+
+package unittests
+
+import (
+	"path"
+	"testing"
+
+	"github.com/google/uuid"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	etcdserver "k8s.io/apiserver/pkg/storage/etcd3/testserver"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/util/cert"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+)
+
+type TestAPIServer struct {
+	ClientSet      clientset.Interface
+	KubeConfigFile string
+}
+
+func StartAPITestServer(t *testing.T) TestAPIServer {
+	cfg := etcdserver.NewTestConfig(t)
+	etcdClient := etcdserver.RunEtcd(t, cfg)
+	storageConfig := storagebackend.NewDefaultConfig(path.Join(uuid.New().String(), "registry"), nil)
+	storageConfig.Transport.ServerList = etcdClient.Endpoints()
+
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{}, storageConfig)
+	t.Cleanup(server.TearDownFn)
+
+	clientSet := clientset.NewForConfigOrDie(server.ClientConfig)
+
+	kubeConfigFile := writeKubeConfigForWardleServerToKASConnection(t, server.ClientConfig)
+
+	return TestAPIServer{
+		ClientSet:      clientSet,
+		KubeConfigFile: kubeConfigFile,
+	}
+}
+
+func writeKubeConfigForWardleServerToKASConnection(t *testing.T, kubeClientConfig *rest.Config) string {
+	// write a kubeconfig out for starting other API servers with delegated auth.  remember, no in-cluster config
+	// the loopback client config uses a loopback cert with different SNI.  We need to use the "real"
+	// cert, so we'll hope we aren't hacked during a unit test and instead load it from the server we started.
+	wardleToKASKubeClientConfig := rest.CopyConfig(kubeClientConfig)
+
+	servingCerts, _, err := cert.GetServingCertificatesForURL(wardleToKASKubeClientConfig.Host, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	encodedServing, err := cert.EncodeCertificates(servingCerts...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	wardleToKASKubeClientConfig.CAData = encodedServing
+
+	for _, v := range servingCerts {
+		t.Logf("Client: Server public key is %v\n", dynamiccertificates.GetHumanCertDetail(v))
+	}
+	certs, err := cert.ParseCertsPEM(wardleToKASKubeClientConfig.CAData)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, curr := range certs {
+		t.Logf("CA bundle %v\n", dynamiccertificates.GetHumanCertDetail(curr))
+	}
+
+	adminKubeConfig := createKubeConfig(wardleToKASKubeClientConfig)
+	tmpDir := t.TempDir()
+	kubeConfigFile := path.Join(tmpDir, "kube.config")
+	if err := clientcmd.WriteToFile(*adminKubeConfig, kubeConfigFile); err != nil {
+		t.Fatal(err)
+	}
+
+	return kubeConfigFile
+}
+
+func createKubeConfig(clientCfg *rest.Config) *clientcmdapi.Config {
+	clusterNick := "cluster"
+	userNick := "user"
+	contextNick := "context"
+
+	config := clientcmdapi.NewConfig()
+
+	credentials := clientcmdapi.NewAuthInfo()
+	credentials.Token = clientCfg.BearerToken
```

Comment:
Almost exactly the same `createKubeConfig` function also exists in https://github.com/kubernetes/kubernetes/blob/23790ec7fa8fd3c5b18edd85bc12ba4d76f50e1f/test/e2e/framework/test_context.go#L427-L463.

The only difference is that the other one also sets
```
 	credentials.TokenFile = clientCfg.BearerTokenFile
```

If we keep this helper under test/e2e, then I should move that function into a common location.

Reply:
I removed the function from `test_context.go` and made sure that the copy in the new `test/utils/apiserver` package is identical (i.e. with `credentials.TokenFile = clientCfg.BearerTokenFile`).